src/openvpn/ssl_openssl.c

```diff
@@ -1501,15 +1501,23 @@ tls_ctx_use_management_external_key(struct tls_root_ctx *ctx)
     }
     EVP_PKEY_free(privkey);
 #else  /* ifdef HAVE_XKEY_PROVIDER */
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
     if (EVP_PKEY_id(pkey) == EVP_PKEY_RSA)
+#else /* OPENSSL_VERSION_NUMBER < 0x30000000L */
+    if (EVP_PKEY_is_a(pkey, "RSA"))
+#endif /* OPENSSL_VERSION_NUMBER < 0x30000000L */
     {
         if (!tls_ctx_use_external_rsa_key(ctx, pkey))
         {
             goto cleanup;
         }
     }
 #if (OPENSSL_VERSION_NUMBER > 0x10100000L) && !defined(OPENSSL_NO_EC)
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
     else if (EVP_PKEY_id(pkey) == EVP_PKEY_EC)
+#else /* OPENSSL_VERSION_NUMBER < 0x30000000L */
+    else if (EVP_PKEY_is_a(pkey, "EC"))
+#endif /* OPENSSL_VERSION_NUMBER < 0x30000000L */
```

Comment:
These two chunks might not be really that important. They are inside the block if we do not have `HAVE_XKEY_PROVIDER`. Then instead of implementing external key usage by using a provider (xkey) we use EC_KEY_METHOD and RSA_KEY_METHOD method and a ton of other deprecated APIs.

Reply:
I don't have a strong preference. You could add a note that this code path is only used when we don't use xkey provider, ie only when the depracted *_KEY_METHOD are used. But I will accept the patch either way.